utils/probing.py

```diff
@@ -0,0 +1,211 @@
+        scaled_intercept = logistic_regression_model.intercept_
+
+        # Unscale the coefficients
+        probe_coef = scaled_coef / std
+
+        # Unscale the intercept
+        probe_intercept = scaled_intercept - (scaled_coef * mean / std).sum(axis=1)
+    else:
+        logistic_regression_model = pipeline.named_steps['logisticregression']
+        probe_coef = logistic_regression_model.coef_
+        probe_intercept = logistic_regression_model.intercept_
+
+    return probe_coef, probe_intercept
+
+
+def cache_and_save_probe_coefficients(
+    model: HookedTransformer,
+    act_folder: str,
+    save_folder: str,
+    labels: Int[np.ndarray, "batch"],
+    component: str,
+    seq_len: int,
+    with_scaler: bool = True,
+    max_iter: int = 100,
+):
+    """
+    Cache and save probe coefficients for all layers and positions.
+    """
+
+    probe_coefficients = torch.zeros((model.cfg.n_layers, seq_len, model.cfg.d_model))
+    probe_intercepts = torch.zeros((model.cfg.n_layers, seq_len))
+    probe_scores = torch.zeros((model.cfg.n_layers, seq_len))
+
+    for layer in range(model.cfg.n_layers):
+        for pos in range(seq_len):
+            probe, score = train_probe_at_layer_pos(
+                model=model, 
+                act_folder=act_folder, 
+                labels=labels, 
+                layer=layer, 
+                pos=pos, 
+                component=component,
+                with_scaler=with_scaler,
+                max_iter=max_iter,
+            )
+
+            probe_coef, probe_intercept = get_probe_direction(probe, with_scaler)
+            
+            probe_coefficients[layer, pos] = torch.tensor(probe_coef)
+            probe_intercepts[layer, pos] = torch.tensor(probe_intercept)
+            probe_scores[layer, pos] = torch.tensor(score)
+
+    # save as pickle file
+    probe_coefficients_path = os.path.join(save_folder, f"probe_coefficients_{component}.pkl")
+    probe_intercepts_path = os.path.join(save_folder, f"probe_intercepts_{component}.pkl")
+    probe_scores_path = os.path.join(save_folder, f"probe_scores_{component}.pkl")
+    
+    with open(probe_coefficients_path, "wb") as f:
+        pickle.dump(probe_coefficients, f)
+    with open(probe_intercepts_path, "wb") as f:
+        pickle.dump(probe_intercepts, f)
+    with open(probe_scores_path, "wb") as f:
+        pickle.dump(probe_scores, f)
+
+    return probe_coefficients, probe_intercepts, probe_scores
+
+
+def apply_probe_to_cache(
+    model: HookedTransformer,
+    cache: ActivationCache,
+    component: str,
+    probe_coefs: Float[Tensor, "layer pos d_model"],
+    alpha: float = 1.0,
+    device: str = "cuda",
+    verbose: bool = False,
+) -> ActivationCache:
+    """
+    Train a probe on a given part of the model, apply the direction of the probe to the cache, and return the new cache.
+    """
+    new_cache = copy.deepcopy(cache)
+    n_layers = model.cfg.n_layers
+    seq_len = cache["blocks.0.hook_mlp_out"].shape[1]
```

Comment:
Minor: I think it's slightly bad practice to assume that the cache has a certain random key. Could you get the seq len using `component` as the key?

Reply:
That should work--I've added the component as the key